app/src/main/scala/org/alephium/explorer/persistence/queries/HashrateQueries.scala

```diff
@@ -33,33 +34,42 @@ trait HashrateQueries extends CustomTypes {
   def getHashratesQuery(from: TimeStamp,
                         to: TimeStamp,
                         interval: Int): DBActionSR[(TimeStamp, BigDecimal)] = {
-    val fromInstant = Instant.ofEpochMilli(from.millis)
-    val toInstant   = Instant.ofEpochMilli(to.millis)
 
     sql"""
         SELECT timestamp, value
         FROM hashrates
-        WHERE interval_type = #$interval
-        AND timestamp >= '#${fromInstant.toString}'
-        AND timestamp <= '#${toInstant.toString}'
+        WHERE interval_type = $interval
+        AND timestamp >= $from
+        AND timestamp <= $to
         ORDER BY timestamp
       """.as[(TimeStamp, BigDecimal)]
   }
 
   def computeHashratesAndInsert(from: TimeStamp, intervalType: Int): DBActionW[Int] = {
-    val hashrates = (intervalType match {
-      case 0 => compute10MinutesHashrateRawString(from)
-      case 1 => computeHourlyHashrateRawString(from)
-      case 2 => computeDailyHashrateRawString(from)
-      case _ => "()"
-    })
-
-    sqlu"""
-      INSERT INTO hashrates (timestamp, value, interval_type)
-      (#$hashrates)
-      ON CONFLICT (timestamp, interval_type) DO UPDATE
-      SET value = EXCLUDED.value
-    """
+    if (intervalType >= 0 || intervalType <= 2) {
+      val dateGroup = (intervalType match {
+        case 0 => tenMinutesQuery
+        case 1 => hourlyQuery
+        case 2 => dailyQuery
+        case _ => ""
+      })
+
+      sqlu"""
+        INSERT INTO hashrates (timestamp, value, interval_type)
+        SELECT
+        EXTRACT(EPOCH FROM ((#$dateGroup) AT TIME ZONE 'UTC')) * 1000 as ts,
+        AVG(hashrate),
+        $intervalType
+        FROM block_headers
+        WHERE timestamp >= $from
```

Comment:
shall the condition be `timestamp >= $from && timestamp < ${from + interval}`?

Reply:
no, the `from` here is to avoid re-computing  hashrates every time from the first block. Before running the request we check when was the latest hashrates computed, we step back and start from that.

Reply:
Is it possible that this query will return a huge number of elements?

Reply:
easy to compute, if you need to compute from scratch the 10 minutes interval and let say Alephium run since 1 year:
`6 * 24 * 365 = 52560`
but the "from scratch" case is computed only once, as we then to what I described just above.

Reply:
sounds good to me!